<a href="https://colab.research.google.com/github/gurjan/pyda/blob/main/math_for_data_analysis/hw1_math.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import sympy as sym

Представим фильмы в виде бинарных векторов (numpy array) используя следующие признаки:

[Комедия, Боевик, Исторический, Бред Питт, Бенедикт Камбербэтч, Квентин Тарантино]



In [ ]:
film_1 = np.array([0,0,0,0,1,0])      # Игра в имитацию

film_2 = np.array([0,1,0,1,0,0])      # Ярость

film_3 = np.array([0,0,1,1,1,0])      # 12 лет рабства

film_4 = np.array([1,0,0,1,0,1])      # Однажды в Голливуде

# Уровень 0:
## Задание 1
Найти угол между векторами-фильмами "Ярость" и "12 лет рабства". Результатом будет являться косинусная мера, которую можно использовать для определения похожести векторов.

In [ ]:
# скалярное произведение (сумм покоординатных сложений)
np.dot(film_2, film_3)

1

In [ ]:
# найдем угол между векторами исходя из определения скалярного произведения 
# для начала находим Евклидову норму для каждого из векторов
np.linalg.norm(film_2)*np.linalg.norm(film_3)

2.4494897427831783

In [ ]:
# делим скарярное произведение на Евклидову норму, получаем косинус угла между векторами
np.dot(film_2, film_3)/(np.linalg.norm(film_2)*np.linalg.norm(film_3))

0.40824829046386296

In [ ]:
# косинус ковертируем в величину угла - в радианах 
np.arccos(np.dot(film_2, film_3)/(np.linalg.norm(film_2)*np.linalg.norm(film_3)))

1.1502619915109316

In [ ]:
# ковертируем в градусы
np.rad2deg(np.arccos(np.dot(film_2, film_3)/(np.linalg.norm(film_2)*np.linalg.norm(film_3))))

65.90515744788931

Угол между векторами равен 65.91 градусам

# Уровень 1:
## Задание 2
Вычислить с помощью Python значение первой производной для функции  в точке x=1

In [ ]:
# определяем какие параметры являются аналитическими
x=sym.Symbol('x')

In [ ]:
3*x*(x**2+x)**3

3*x*(x**2 + x)**3

In [ ]:
# получаем аналитическую призводную
res = sym.diff(3*x*(x**2+x)**3,x)

In [ ]:
res.subs(x,1)

132

##Задание 3
Объединить векторы фильмов в матрицу, где каждый вектор является строкой матрицы. Умножьте вектор [1,2,3,4] на полученную матрицу

In [ ]:
films = np.stack((film_1, film_2, film_3, film_4))

In [ ]:
data = np.array([1, 2, 3, 4]) 

In [ ]:
res = np.dot(data, films)
res

array([4, 2, 3, 9, 4, 4])